### PR TITLE
Improve performance join decomposition

### PR DESCRIPTION
**Pull request by derekkraan**
Should hopefully fix #15

### CODE REVIEW

File: lib/delta_crdt/aw_lww_map.ex [hunk 1]
```diff
@@ -118,8 +122,18 @@ defmodule DeltaCrdt.AWLWWMap do
     end)
   end
 
+  defp all_dots(%{state: state}) do
+    Enum.reduce(state, MapSet.new(), fn {_key, dot_map}, dots ->
```
Comment (sb8244) [inline lib/delta_crdt/aw_lww_map.ex:126]: I believe that this function may be the performance bottleneck on add:

image

**Comment by sb8244** [comment]
While the reads seem much faster, the adds appear to be O(n^2) now. I'm at about 10 minutes for 27k adds.

**Comment by sb8244** [comment]
Looking through the code, I wonder if it's possible to keep `all_dots` in the top level CausalDotMap. That would allow for instant access to all of the dots, at the cost of memory. I'm sure it's more complex than that though 🙃

**Comment by derekkraan** [comment]
I'm closing this in favour of #16

**Pull request closed without merging**